diff --git a/Setup.hs b/Setup.hs
index e3e766f..bf2fd68 100644
--- a/Setup.hs
+++ b/Setup.hs
@@ -12,6 +12,7 @@ import Distribution.Simple.PreProcess
 import Distribution.Simple.Program
 import Distribution.Simple.Setup hiding (Flag)
 import Distribution.System
+import Distribution.Utils.Path (makeSymbolicPath)
 import System.Environment
 
 #ifdef MIN_VERSION_Cabal
@@ -144,8 +145,8 @@ main = do
               }
            }
           configFlags' = confFlags {
-            configExtraLibDirs = libDirs ++ configExtraLibDirs confFlags,
-            configExtraIncludeDirs = includeDirs ++ configExtraIncludeDirs confFlags
+            configExtraLibDirs = (map makeSymbolicPath libDirs) ++ configExtraLibDirs confFlags,
+            configExtraIncludeDirs = (map makeSymbolicPath includeDirs) ++ configExtraIncludeDirs confFlags
            }
       confHook simpleUserHooks (genericPackageDescription', hookedBuildInfo) configFlags'
    }
